### PR TITLE
Tween Serialization + Editor Support

### PR DESCRIPTION
**Pull request by kkjamie**
# Setup your tweens in the editor!

## No Code Required
You can now set up tweens in the editor, using the "TweenBuilder" script, it will allow you to choose which type of tween to create, and then select all the parameters you want, and it will be saved in the scene/prefab!
tween-builder

## ScriptableObject support
Just tick "Use referenced Config" and you can configure a tween in a scriptable object. and apply that tween to several objects. (The target cannot be serialized in a scriptable object, so it will just use the TweenBuilder object as a target or look for  the correct component on there).

scriptable-object

## Supports Most Tweens!
it supports most of the built in tween types, and can be made most any of the constructors and extension functions you already use*
tween-creator-functions
\* We filter out functions that have unsupported args; those not supported by ArgsList, such as quaternion. However we also make special cases like `Func<float,float>`. Although we cannot serialize this, we know that this type is the easing function, therefore we actually just store it as a string, and we can lookup the key later

## Extendable
The list above comes from a public static dictionary, so you will be able to register your own creator functions by using `TweenCreatorFunctionStore` (API example incoming)

## Easings
easing-functions
We support all built in easings and again you can plug your own in using `EasingFunctionsStore.EasingFunctions.Add("key", MyEasingFunction)`

# Current State

## TODO

- Add more built in creator functions so we can move between 2 referenced transforms (won't work for scriptable objects)
- Add support for adding your own Tween Creation Functions
- Some improved error handling in a few places
- Draw warnings to inspector about null targets (and explain the behaviour in this instance)
- Add support for handles when specifying Vector 3 positions
- Looping support

## What can it not do?
Sequenced animations are not possible, I thought about how to implement it but it gets pretty complicated, at that point I'd say you want something really custom anyway, and may as well jsut write the code, This is intended for quickly adding basic effects like popping/fading etc... Complicated chains are probably better to be explicitly expressed as code.

### CODE REVIEW

File: Tween/Serialization/TweenBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+		public void Play()
+		{
+			var config = useReferencedConfig ? tweenConfigScriptableObject.Config : TweenConfig;
+			var tween = BuildTween(config, gameObject);
+			tween.Play();
+		}
+
+#if UNITY_EDITOR
+		public bool PlayOnStart
```
Comment (LukeAvery92) [inline Tween/Serialization/TweenBuilder.cs]: Why should this not be accessible outside editor time?

Comment (LukeAvery92) [reply Tween/Serialization/TweenBuilder.cs]: No i wouldn't have but wanted to understand where it was being used but if only on a inspector class then I don't mind.

**Comment by Rollo-Liu** [comment]
Not the pro-dark-skin! Hummmmmmm
:trollface:

File: Tween/Serialization/TweenConfigScriptableObject.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using UnityEngine;
+
+namespace DUCK.Tween.Serialization
+{
+	[CreateAssetMenu(menuName = "Tween Config (Fancy™)", order = 230)]
```
Comment (Rollo-Liu) [inline Tween/Serialization/TweenConfigScriptableObject.cs]: Oh YES.

**Comment by kkjamie** [comment]
@Rollo-Liu  Haha, I suggest you checkout the branch and have a play!

**Comment by kkjamie** [comment]
I just completed all the TODOs from the code and the description above apart from 1 Adding gizmos/handles for manipulating vecrtor3s in a meaningful way. This is a nice to have, and would take a lot more work. I would rather just get this merged in.

File: Tween/Easings/EasingFunctionStore.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+using System;
+using System.Collections.Generic;
+
+namespace DUCK.Tween.Easings
+{
+	public class EasingFunctionStore
+	{
+		public static Dictionary<string, Func<float, float>> EasingFunctions { get; set; }
+		static EasingFunctionStore()
```
Comment (AndrewThompson85) [inline Tween/Easings/EasingFunctionStore.cs]: Does this have to be a method? could it not be a backing field for the property?

File: Tween/Easings/EasingFunctionStore.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+using System;
+using System.Collections.Generic;
+
+namespace DUCK.Tween.Easings
+{
+	public class EasingFunctionStore
+	{
+		public static Dictionary<string, Func<float, float>> EasingFunctions { get; set; }
```
Comment (AndrewThompson85) [inline Tween/Easings/EasingFunctionStore.cs]: Does this dictionary need to have a public setter ?

Comment (kkjamie) [reply Tween/Easings/EasingFunctionStore.cs]: Yeah it should not

**Pull request merged**